frontend/public/components/utils/resource-log.jsx

```diff
@@ -130,15 +133,21 @@ export class ResourceLog extends SafetyFirst {
   // Handler for websocket onopen event
   _onOpen(){
     this._buffer = [];
-    this.setState({status: 'streaming'});
+    this._updateStatus('streaming');
   }
 
   // Push one line to the buffer. First item is removed if buffer is at limit.
   _pushToBuffer(text){
-    if (this._buffer.length === this.props.bufferSize) {
-      this._buffer.shift();
+    const prev = this.state.currentLine;
```

Comment:
Considering suggesting some small helper functions, like `if(hasNewLine(text))` `if(bufferMaxReached())` just to help charify this logic.

A little confusing why `_pushToBuffer()` also is responsible for returning either `''` or `next`. 
`123` has `const currentLine = this._pushToBuffer(text);`
Wondering if it shouldn't be:

```JavaScript 
this._pushToBuffer(text); // or updateBuffer(), as it might not actually "push" according to the logic.
const currentLine = lastBufferLine();
```

Reply:
You have enough bookkeeping for the `Buffer` that it could potentially be extracted out into its own component/class with its own methods.

Reply:
Potentially you could extract a `Streamer` `class` as well, depending on how far you wanted to refactor (just for separation of concerns).

Reply:
Yeah, I removed the LineBuffer class that existed before because it was kind of a mess and was only used in one place, but it might make sense to wrap this buffer in a class and have it keep track of some internal state on it's own.

Reply:
This would probably be a follow on. Maybe we should create a Jira issue so we don't lose track of it.